src/radical/pilot/agent/executing/popen.py

```diff
@@ -135,12 +135,26 @@ def work(self, units):
     def _handle_unit(self, cu):
 
         try:
+            descr = cu['description']
+
+            # ensure that the named env exists
+            env = descr.get('named_env')
+            if env:
+                if not os.path.isdir('%s/%s' % (self._pwd, env)):
+                    raise ValueError('invalid named env %s for task %s' 
+                                    % (env, cu['uid']))
+                pre = ru.as_list(descr.get('pre_exec'))
+                pre.insert(0, '. %s/%s/bin/activate' % (self._pwd, env))
+                pre.insert(0, '. %s/deactivate'      % (self._pwd))
```

Comment:
do we create an executable file `deactivate`? otherwise it should be `pre.insert(0, 'deactivate')`

p.s. for `conda` env it would be `pre.insert(0, 'conda %s/%s/bin/deactivate' % (self._pwd, env))`, but we keep it for later

Reply:
The bootstrapper does indeed create a `deactivate` script [here](https://github.com/radical-cybertools/radical.pilot/blob/devel/src/radical/pilot/agent/bootstrap_0.sh#L22).  That is a bit simplistic at the moment and will get some more sophistication eventually - but this is what is being sourced here.